### PR TITLE
refactor: expose shared recipe builders from components

### PR DESCRIPTION
**Pull request by HuiyingLi**
## Summary

- Move duplicated LLM/VLM builder helpers into component-level modules for checkpoint config, distributed setup, loss construction, optimizer construction, LR scheduler construction, step scheduler construction, and wandb setup.
- Export the new public builder paths from the relevant `components/*/__init__.py` files.
- Keep recipe-level backwards compatibility by importing the moved helpers into the existing recipe modules.
- Update sibling recipes to consume the new component paths where applicable, including the retrieval mining distributed setup.
- Update Dion optimizer branch tests to exercise the shared component optimizer builder directly.

## Validation

- `uv run ruff format ...` on touched code files
- `uv run ruff check ...` on touched code files
- Component builder import smoke check
- Shared `build_optimizer` smoke check
- `python -m py_compile ...` on touched Python modules
- `uv run pytest tests/unit_tests/optim/test_dion_optimizer_utils.py::TestBuildOptimizerDionBranch`
- `python -m py_compile nemo_automodel/recipes/retrieval/mine_hard_negatives.py`
- `git diff --check`

Earlier focused recipe pytest collection was attempted, but this environment fails before test execution due to existing binary/import issues: missing `torchvision::nms` and a Transformer Engine undefined symbol.

### CODE REVIEW

File: nemo_automodel/components/checkpoint/api.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
```
Comment (akoumpa) [inline nemo_automodel/components/checkpoint/api.py]: Hi @HuiyingLi, I'm proposing the following standardization for each component:
- config.py
- api.py
these are the only two files anyone would have to look at to figure out how to use a component. However, if that's not possible, or the interface becomes too convoluted, then we may also want to consider applying this structure in a hierarchical fashion.

Also a reminder, we want this to be 100% pythonic, no yamls at the component level. To confirm this is working in practice one can inspect the code to search for configNodes or Any types and also count the rate of getattr function is used -- pythonic objects need not use getattr.

File: nemo_automodel/components/distributed/init_utils.py [hunk 2]
```diff
@@ -161,3 +162,17 @@ def destroy_global_state():
     if torch.distributed.is_available() and torch.distributed.is_initialized():
         torch.distributed.destroy_process_group()
     signal.signal(signal.SIGINT, signal.SIG_DFL)
+
+
+def build_distributed(cfg_dist: dict[str, Any]) -> DistInfo:
```
Comment (akoumpa) [inline nemo_automodel/components/distributed/init_utils.py]: I'm wondering if we even need this function or instead we could call initialize_distributed directly?

Comment (HuiyingLi) [reply nemo_automodel/components/distributed/init_utils.py]: sure

File: nemo_automodel/components/optim/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    weight_decay: float = 0.0
+    mu: float = 0.95
+    betas: tuple[float, float] = (0.9, 0.95)
+    epsilon: float = 1e-8
+    adjust_lr: str = "spectral_norm"
+    scalar_opt: str = "adamw"
+    scalar_betas: tuple[float, float] = (0.9, 0.999)
+    scalar_eps: float = 1e-8
+
+    def build_dion(self, model: torch.nn.Module, device_mesh: DeviceMesh | None) -> torch.optim.Optimizer:
```
Comment (akoumpa) [inline nemo_automodel/components/optim/optimizer.py]: Hi @HuiyingLi is there a way to avoid the specific `build_dion` and instead rely on a `build`?

File: nemo_automodel/components/optim/optimizer.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        return _fully_shard_megatron_optimizer(model_part, optimizer)
+    return optimizer
+
+
+# ---------------------------------------------------------------------------
+# Builders
+# ---------------------------------------------------------------------------
+
+
+def build_optimizer(
```
Comment (akoumpa) [inline nemo_automodel/components/optim/optimizer.py:614]: @HuiyingLi can we eliminate this and instead rely on the .build of the configs?

Comment (akoumpa) [reply nemo_automodel/components/optim/optimizer.py:614]:
```
def build_optimizer(name, params, **kwargs):
    return {'adam': AdamConfig, 'sgd': SGDConfig}[name](**kwargs).build(params)
```
if not possible to eliminate, please consider using the above, so that we have everything typed in the configs, but allow a flexible init function.

File: nemo_automodel/components/optim/optimizer.py [hunk 5]
```diff
@@ -184,25 +215,210 @@ class MuonConfig(OptimizerConfig):
...
+                kwargs[attr] = dtype_from_str(val)
+        if foreach is not None:
+            kwargs.setdefault("foreach", foreach)
+
+        optimizers: list[torch.optim.Optimizer] = []
+        for part in getattr(model, "parts", [model]):
+            trainable_params = [p for p in part.parameters() if p.requires_grad]
+            assert len(trainable_params) > 0, "trainable_params cannot be empty"
+            opt = self.factory(params=trainable_params, **kwargs)
+            opt = maybe_shard_optimizer(part, opt, distributed_config, allow=True)
```
Comment (akoumpa) [inline nemo_automodel/components/optim/optimizer.py]: for the `maybe_shard_optimizer` I'm not sure it should be in the optim component.

File: nemo_automodel/recipes/_typed_config.py [hunk 6]
```diff
@@ -186,15 +167,17 @@ def step_scheduler(self) -> StepSchedulerConfig:
...
+
         node = self._raw.get("optimizer", None)
         if node is None:
             return None
         factory, kwargs = _callable_and_kwargs(node)
-        return OptimizerSpec(factory, kwargs)
+        return build_optimizer_config(factory, kwargs)
 
     @cached_property
     def loss_fn(self) -> LossSpec | None:
```
Comment (HuiyingLi) [inline nemo_automodel/recipes/_typed_config.py]: @akoumpa  do you intend to keep LossSpec?

**Comment by jgerh** [comment]
Found some possible MDX/JSX block component (Info, Tip, Note, Accordion) syntax issues with fern/versions/nightly/pages/guides/llm/finetune.mdx and fern/versions/v0.4/pages/guides/llm/finetune.mdx. Tech pubs is investigating.

**Review by oliverholworthy — APPROVED** [review]
Reviewed from the retrieval models side. no blocker from my side.

File: nemo_automodel/recipes/_dist_setup.py [hunk 7]
```diff
@@ -243,3 +243,18 @@ def setup_distributed(cfg: Any, world_size: Optional[int] = None) -> MeshContext
         device_mesh=device_mesh,
         moe_mesh=moe_mesh,
     )
+
+
+def shard_optimizers_for_megatron_fsdp(model, optimizers, distributed_config, *, allow=True):
```
Comment (akoumpa) [inline nemo_automodel/recipes/_dist_setup.py:248]: this is not great, but not a catastrophe either.

**Comment by akoumpa** [comment]
I've reviewed the last 5 commits, this is good to go from my side @HuiyingLi

**Review by akoumpa — APPROVED** [review]
